cache/contenthash/checksum.go

```diff
@@ -506,21 +506,77 @@ func (cc *cacheContext) includedPaths(ctx context.Context, m *mount, p string, o
 	txn := cc.tree.Txn()
 	root = txn.Root()
 	var (
-		updated bool
-		iter    *iradix.Iterator
-		k       []byte
-		kOk     bool
+		updated        bool
+		iter           *iradix.Iterator
+		k              []byte
+		kOk            bool
+		origPrefix     string
+		resolvedPrefix string
 	)
 
 	iter = root.Iterator()
 
 	if opts.Wildcard {
-		k, _, kOk = iter.Next()
+		// For consistency with what the copy implementation in fsutil
+		// does: split pattern into non-wildcard prefix and rest of
+		// pattern, then follow symlinks when resolving the non-wildcard
+		// prefix.
+
+		d1, d2 := splitWildcards(p)
+		if d1 != "/" {
+			origPrefix = d1
+			k = convertPathToKey([]byte(d1))
+			linksWalked := 0
+			if d2 != "" {
+				// getFollowLinks only handles symlinks in path
+				// components before the last component, so
+				// handle last component in d1 specially.
+				for {
+					v, ok := root.Get(k)
```

Comment:
Not sure I get this. Isn't this supposed to walk per component? see `getFollowLinksWalk()`

Reply:
There is a call to `getFollowLinks` below (line 568) which does the walk by component. This loop is just to handle the case where the final component is a symlink, which `getFollowLinks` does not handle.

I've added a test case for this specific scenario.